cachi2/core/package_managers/bundler/parser.py

```diff
@@ -0,0 +1,160 @@
+import json
+import logging
+import re
+import subprocess
+from importlib.resources import files
+from typing import Optional, Union
+
+import pydantic
+
+from cachi2.core.errors import PackageRejected, UnexpectedFormat
+from cachi2.core.rooted_path import PathOutsideRoot, RootedPath
+
+log = logging.getLogger(__name__)
+
+GEMFILE = "Gemfile"
+GEMFILE_LOCK = "Gemfile.lock"
+
+
+class _GemMetadata(pydantic.BaseModel):
+    """
+    Base class for gem metadata.
+
+    Attributes:
+        name:       The name of the gem.
+        version:    The version of the gem.
+    """
+
+    name: str
+    version: str
+
+
+class GemDependency(_GemMetadata):
+    """
+    Represents a gem dependency.
+
+    Attributes:
+        source:     The source URL of the gem.
+        checksum:   The checksum of the gem.
+    """
+
+    source: str
+    checksum: Optional[str] = None
+
+    @property
+    def purl(self) -> str:
+        """Get purl for this gem dependency."""
+        return NotImplemented
+
+
+class GitDependency(_GemMetadata):
+    """
+    Represents a git dependency.
+
+    Attributes:
+        url:        The URL of the git repository.
+        revision:   Commit hash.
+        branch:     The branch of the git repository.
+    """
+
+    url: str
+    revision: str
+    branch: Optional[str] = None
+
+    @property
+    def purl(self) -> str:
+        """Get purl for this git dependency."""
+        return NotImplemented
+
+    @pydantic.field_validator("url")
+    def validate_git_url(cls, url: str) -> str:
+        """Validate that the git URL is a valid git URL."""
+        if not url.startswith("https://"):
+            raise ValueError("GIT dependencies must use HTTPS protocol")
```

Comment:
Is this all we want to validate? I guess the Ruby lockfile parser would flag incorrect URLs, but if we really perform validation, it might be worth using our `extract_git_info` which parses the whole URL into bits and which we use elsewhere.

Reply:
I believe `urlparse` is enough here. `extract_git_info` needs VCS URL.
```
>>> from urllib.parse import urlparse
>>> parsed = urlparse("https://github.com/3scale/json-schema.git")
>>> parsed
ParseResult(scheme='https', netloc='github.com', path='/3scale/json-schema.git', params='', query='', fragment='')
```

Reply:
Right, it was because of `urlparse` I suggested it in the first place so works for me.